core/drivers/crypto/hisilicon/hisi_qm.c

```diff
@@ -0,0 +1,791 @@
+	QM_MB_CMD_STOP_QM,
+};
+
+struct qm_mailbox {
+	union {
+		struct {
+			uint16_t w0;
+			uint16_t queue;
+			unsigned int base_l;
+			unsigned int base_h;
+			unsigned int token;
+		};
+		uint64_t x[2];
+	};
+};
+
+struct qm_dfx_registers {
+	const char *reg_name;
+	unsigned int reg_offset;
+};
+
+static const struct qm_dfx_registers qm_dfx_regs[] = {
+	{ .reg_name = "QM_ECC_1BIT_CNT           ", .reg_offset = 0x104000 },
+	{ .reg_name = "QM_ECC_MBIT_CNT           ", .reg_offset = 0x104008 },
+	{ .reg_name = "QM_DFX_MB_CNT             ", .reg_offset = 0x104018 },
+	{ .reg_name = "QM_DFX_DB_CNT             ", .reg_offset = 0x104028 },
+	{ .reg_name = "QM_DFX_SQE_CNT            ", .reg_offset = 0x104038 },
+	{ .reg_name = "QM_DFX_CQE_CNT            ", .reg_offset = 0x104048 },
+	{ .reg_name = "QM_DFX_SEND_SQE_TO_ACC_CNT", .reg_offset = 0x104050 },
+	{ .reg_name = "QM_DFX_WB_SQE_FROM_ACC_CNT", .reg_offset = 0x104058 },
+	{ .reg_name = "QM_DFX_ACC_FINISH_CNT     ", .reg_offset = 0x104060 },
+	{ .reg_name = "QM_DFX_CQE_ERR_CNT        ", .reg_offset = 0x1040b4 },
+	{ .reg_name =  NULL, 0 }
+};
+
+void hisi_qm_get_version(struct hisi_qm *qm)
+{
+	qm->version = io_read32(qm->io_base + HISI_QM_REVISON_ID_BASE) &
+		      HISI_QM_REVISON_ID_MASK;
+}
+
+static void qm_db(struct hisi_qm *qm, uint16_t qn, uint8_t cmd, uint16_t index,
+		  uint8_t priority)
+{
+	uint64_t doorbell = 0;
+
+	doorbell = qn | SHIFT_U64(cmd, QM_DB_CMD_SHIFT) |
+		   SHIFT_U64(QM_DB_RAND_DATA, QM_DB_RAND_DATA_SHIFT) |
+		   SHIFT_U64(index, QM_DB_INDEX_SHIFT) |
+		   SHIFT_U64(priority, QM_DB_PRIORITY_SHIFT);
+
+	io_write64(qm->io_base + QM_DOORBELL_SQ_CQ_BASE, doorbell);
+}
+
+static enum hisi_drv_status qm_wait_mb_ready(struct hisi_qm *qm)
+{
+	unsigned int val = 0;
+
+	/* return 0 mailbox ready, HISI_QM_DRVCRYPT_ETMOUT hardware timeout */
+	if (IO_READ32_POLL_TIMEOUT(qm->io_base + QM_MAILBOX_BASE, val,
+				   !(val & QM_MB_BUSY_BIT), POLL_PERIOD,
+				   POLL_TIMEOUT)) {
+		return HISI_QM_DRVCRYPT_ETMOUT;
+	}
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static void qm_mb_write(struct hisi_qm *qm, struct qm_mailbox *mb)
+{
+	vaddr_t dst = qm->io_base + QM_MAILBOX_BASE;
+
+	write_64bit_pair(dst, mb->x[0], mb->x[1]);
+}
+
+static enum hisi_drv_status qm_mb(struct hisi_qm *qm, uint8_t cmd,
+				  vaddr_t dma_addr, uint16_t qn, uint8_t op)
+{
+	struct qm_mailbox mb = { };
+
+	mb.w0 = cmd | SHIFT_U32(op, QM_MB_OP_SHIFT) |
+		BIT32(QM_MB_BUSY_SHIFT);
+	mb.queue = qn;
+	reg_pair_from_64(dma_addr, &mb.base_h, &mb.base_l);
+	mb.token = 0;
+
+	if (qm_wait_mb_ready(qm)) {
+		EMSG("QM mailbox is busy");
+		return HISI_QM_DRVCRYPT_EBUSY;
+	}
+
+	qm_mb_write(qm, &mb);
+
+	if (qm_wait_mb_ready(qm)) {
+		EMSG("QM mailbox operation timeout");
+		return HISI_QM_DRVCRYPT_EBUSY;
+	}
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static void qm_cfg_vft_data(struct hisi_qm *qm, uint8_t vft_type,
+			    unsigned int base, unsigned int number)
+{
+	unsigned int data_h = 0;
+	unsigned int data_l = 0;
```

Comment:
here `data_l` and `data_h` should be `uint32_t`. They are used a `io_write32()` argument.